In [3]:
import pandas as pd
import numpy as np
import cv2
import os
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
from tensorflow.keras.utils import to_categorical
from sklearn.model_selection import train_test_split

# Veriyi oku
labels_df = pd.read_csv("train.csv")
images = []
labels = []

for _, row in labels_df.iterrows():
    img = cv2.imread(os.path.join("train", row["id"]))
    img = cv2.resize(img, (32, 32))
    images.append(img)
    labels.append(row["has_cactus"])

X = np.array(images) / 255.0
y = to_categorical(np.array(labels))

# Veriyi böl
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)),
    MaxPooling2D(2, 2),
    Conv2D(64, (3, 3), activation='relu'),
    MaxPooling2D(2, 2),
    Flatten(),
    Dense(64, activation='relu'),
    Dropout(0.5),
    Dense(2, activation='softmax')
])

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, epochs=5, validation_split=0.1)

# Modeli kaydet
model.save("cactus_model.h5")


C:\ProgramData\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
394/394 ━━━━━━━━━━━━━━━━━━━━ 5s 10ms/step - accuracy: 0.8596 - loss: 0.3108 - val_accuracy: 0.9336 - val_loss: 0.1838
Epoch 2/5
394/394 ━━━━━━━━━━━━━━━━━━━━ 4s 9ms/step - accuracy: 0.9420 - loss: 0.1549 - val_accuracy: 0.9514 - val_loss: 0.1329
Epoch 3/5
394/394 ━━━━━━━━━━━━━━━━━━━━ 4s 10ms/step - accuracy: 0.9530 - loss: 0.1255 - val_accuracy: 0.9614 - val_loss: 0.0976
Epoch 4/5
394/394 ━━━━━━━━━━━━━━━━━━━━ 4s 10ms/step - accuracy: 0.9648 - loss: 0.1021 - val_accuracy: 0.9700 - val_loss: 0.0754
Epoch 5/5
394/394 ━━━━━━━━━━━━━━━━━━━━ 4s 10ms/step - accuracy: 0.9713 - loss: 0.0828 - val_accuracy: 0.9593 - val_loss: 0.1042


Sonuçlar:
Eğitim Doğruluğu: %97.13

Doğrulama (Validation) Doğruluğu: %95.93

Kayıp (Loss): 0.1042
Bu değerler, modelin görüntü sınıflandırmada oldukça iyi performans gösterdiğini gösteriyor.